getline.ts/src/blockchain.ts

```diff
@@ -0,0 +1,237 @@
+     * @returns Block number.
+     */
+    currentBlock(): Promise<BigNumber>;
+    /**
+     * Retrieves a handle to a contract at an address on the blockchain.
+     * @param name Name of contract on the metabackend.
+     * @param address Address of contract on the blockchain.
+     * @returns Contract at this address.
+     */
+    existing(name: string, address: Address): Promise<Contract>;
+    /**
+     * Deploys a new contract on the blockchain.
+     * @param name Name of contract on the metabackend.
+     * @param params Constructor parameters of the contract.
+     * @returns Newly deployed contract.
+     */
+    deploy(name: string, ...params: Array<any>): Promise<Contract>;
+    /**
+     * Returns coinbase of the blockchain client.
+     * @returns Coinbase address.
+     */
+    coinbase(): Promise<Address>;
+}
+
+/**
+ * A contract deployed on a blockchain.
+ */
+export class Contract {
+    /**
+     * Address of the contract on the blockchain.
+     */
+    public address: Address
+    /**
+     * Blockchain on which the contract is located.
+     */
+    private blockchain: Blockchain
+    private instance: Web3.ContractInstance
+
+    constructor(blockchain: Blockchain, instance: Web3.ContractInstance) {
+        this.blockchain = blockchain;
+        this.instance = instance;
+        this.address = new Address(this.blockchain, instance.address);
+    }
+
+    /**
+     * Call a contract method.
+     * @param methodName Name of method to call.
+     * @param params Arguments to pass to method.
+     * @returns Result of call.
+     */
+    public async call(methodName: string, ...params: Array<any>): Promise<any> {
+        let method: any = this.instance[methodName];
+        return new Promise<any>((resolve, reject)=>{
+            let opts = {
+                gas: 1000000,
+            }
+            method(...params, opts, (err, object)=>{
+                if (err != undefined) {
+                    reject(err);
+                    return;
+                }
+                resolve(object);
+            });
+        });
+    }
+}
+
+/**
+ * Implementation of the Blockchain interface that uses the Getline metabackend
+ * and web3.
+ */
+export class GetlineBlockchain {
+    private metabackend: MetabackendClient;
+    private network: string;
+    private web3: Web3;
+    private contractDefinitions: Array<pb.Contract> | undefined;
+
+    constructor(metabackend: MetabackendClient, network: string) {
+        this.metabackend = metabackend;
+        this.network = network;
+
+        let provider = new Web3.providers.HttpProvider("http://localhost:8545")
+        if (typeof window !== 'undefined' && typeof window['web3'] !== 'undefined') {
+            console.log("getline.ts: using injected web3")
+            provider = window['web3'].currentProvider;
+        } else {
+            console.log("getline.ts: connecting to node running on localhost")
+        }
+        this.web3 = new Web3(provider);
```

Comment:
Users might want to inject their own web3 with a custom provider. For example proxy or [Infura](infura.io)

Reply:
Done.